plugins/repository-gcs/src/main/java/io/crate/gcs/GCSRepository.java

```diff
@@ -0,0 +1,153 @@
+
+import com.google.auth.oauth2.ServiceAccountCredentials;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+
+/**
+ * Google Cloud Storage implementation of the BlobStoreRepository
+ * <p>
+ * This repository supports the following settings
+ * <dl>
+ * <dt>{@code bucket}</dt><dd>Bucket name</dd>
+ * <dt>{@code base_path}</dt><dd>Base path (blob name prefix) in the bucket</dd>
+ * <dt>{@code private_key_id}</dt><dd>Private key in PKCS 8 format from the google service account credentials.</dd>
+ * <dt>{@code private_key}</dt><dd>Private key from the google service account credentials.</dd>
+ * <dt>{@code client_id}</dt><dd>client id from the google service account credentials.</dd>
+ * <dt>{@code client_email}</dt><dd>client email from the google service account credentials.</dd>
+ * <dt>{@code endpoint}</dt><dd>Endpoint root URL, only used for testing to connect to an alternative storage.</dd>
+ * </dl>
+ */
+public class GCSRepository extends BlobStoreRepository {
+
+    private static final Logger LOGGER = LogManager.getLogger(GCSRepository.class);
+
+    static final Setting<String> BUCKET_SETTING =
+        Setting.simpleString("bucket", Property.NodeScope);
+
+    static final Setting<String> BASE_PATH_SETTING =
+        Setting.simpleString("base_path", "", Property.NodeScope);
+
+    static final Setting<SecureString> PROJECT_ID_SETTING =
+        Setting.maskedString("project_id");
+
+    static final Setting<SecureString> PRIVATE_KEY_ID_SETTING =
+        Setting.maskedString("private_key_id");
+
+    static final Setting<SecureString> PRIVATE_KEY_SETTING =
+        Setting.maskedString("private_key");
+
+    static final Setting<SecureString> CLIENT_EMAIL_SETTING =
+        Setting.maskedString("client_email");
+
+    static final Setting<SecureString> CLIENT_ID_SETTING =
+        Setting.maskedString("client_id");
+
+    static final Setting<String> ENDPOINT_SETTING =
+        Setting.simpleString("endpoint", Property.NodeScope);
+
+    public GCSRepository(RepositoryMetadata metadata,
+                         NamedXContentRegistry namedXContentRegistry,
+                         ClusterService clusterService,
+                         RecoverySettings recoverySettings) {
+        super(metadata, namedXContentRegistry, clusterService,
+            recoverySettings, buildBasePath(metadata));
+    }
+
+    private static BlobPath buildBasePath(RepositoryMetadata metadata) {
+        final String basePath = BASE_PATH_SETTING.get(metadata.settings());
+        return Strings.hasLength(basePath) ?
+            new BlobPath().add(basePath) :
+            BlobPath.cleanPath();
+    }
+
+    @Override
+    protected BlobStore createBlobStore() throws Exception {
+        Settings settings = metadata.settings();
+
+        final Storage storage;
+        if (ENDPOINT_SETTING.exists(settings)) {
+            // This is only for testing purposes
```

Comment:
This doesn't sound like this should remain here?

Reply:
I don't see any other way to make this work. The Google Cloud library does two steps to connect:

- connect by oauth to authenticate the user
- connect to the storage endpoint 

If I want to connect to a fake server with a custom endpoint, the authentication must be skipped to make it work, because I don't want to authenticate against the oauth service from google.

Reply:
See new version, this should be resolved now.